File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/CallbackFactory.java
/*
 * Copyright (c) 2019 - 2022 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package org.forgerock.android.auth.callback;

import org.forgerock.android.auth.Logger;

import java.util.HashMap;
import java.util.Map;

import lombok.Getter;

/**
 * Factory to manage supported {@link Callback}
 */
public class CallbackFactory {

    private static final String TAG = CallbackFactory.class.getSimpleName();
    private static final CallbackFactory INSTANCE = new CallbackFactory();

    @Getter
    private Map<String, Class<? extends Callback>> callbacks = new HashMap<>();

    private CallbackFactory() {
        register(ChoiceCallback.class);
        register(NameCallback.class);
        register(PasswordCallback.class);
        register(StringAttributeInputCallback.class);
        register(NumberAttributeInputCallback.class);
        register(BooleanAttributeInputCallback.class);
        register(ValidatedPasswordCallback.class);
        register(ValidatedUsernameCallback.class);
        register(KbaCreateCallback.class);
        register(TermsAndConditionsCallback.class);
        register(PollingWaitCallback.class);
        register(ConfirmationCallback.class);
        register(TextOutputCallback.class);
        register(SuspendedTextOutputCallback.class);
        register(ReCaptchaCallback.class);
        register(ConsentMappingCallback.class);
        register(HiddenValueCallback.class);
        register(DeviceProfileCallback.class);
        register(MetadataCallback.class);
        register(WebAuthnRegistrationCallback.class);
        register(WebAuthnAuthenticationCallback.class);
        register(SelectIdPCallback.class);
        register(IdPCallback.class);
        register(DeviceBindingCallback.class);
        register(DeviceSigningVerifierCallback.class);
    }

    /**
     * Returns a cached instance {@link CallbackFactory}
     *
     * @return instance of {@link CallbackFactory}
     */
    public static CallbackFactory getInstance() {
        return INSTANCE;
    }

    /**
     * Register new Callback Class
     *
     * @param callback The callback Class
     */
    public void register(Class<? extends Callback> callback) {
        try {
            callbacks.put(getType(callback), callback);
        } catch (Exception e) {
            Logger.error(TAG, e, e.getMessage());
        }
    }

    public String getType(Class<? extends Callback> callback) throws InstantiationException, IllegalAccessException {
        return callback.newInstance().getType();
    }

}

File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/CallbackFactory.kt
/*
 * Copyright (c) 2019 - 2024 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */
package org.forgerock.android.auth.callback

import org.forgerock.android.auth.Logger.Companion.error

/**
 * Factory to manage supported [Callback]
 */
class CallbackFactory private constructor() {
    internal val callbacks: MutableMap<String, Class<out Callback>> = HashMap()

    init {
        register(ChoiceCallback::class.java)
        register(NameCallback::class.java)
        register(PasswordCallback::class.java)
        register(StringAttributeInputCallback::class.java)
        register(NumberAttributeInputCallback::class.java)
        register(BooleanAttributeInputCallback::class.java)
        register(ValidatedPasswordCallback::class.java)
        register(ValidatedUsernameCallback::class.java)
        register(KbaCreateCallback::class.java)
        register(TermsAndConditionsCallback::class.java)
        register(PollingWaitCallback::class.java)
        register(ConfirmationCallback::class.java)
        register(TextOutputCallback::class.java)
        register(SuspendedTextOutputCallback::class.java)
        register(ReCaptchaCallback::class.java)
        register(ConsentMappingCallback::class.java)
        register(HiddenValueCallback::class.java)
        register(DeviceProfileCallback::class.java)
        register(MetadataCallback::class.java)
        register(WebAuthnRegistrationCallback::class.java)
        register(WebAuthnAuthenticationCallback::class.java)
        register(SelectIdPCallback::class.java)
        register(IdPCallback::class.java)
        register(DeviceBindingCallback::class.java)
        register(DeviceSigningVerifierCallback::class.java)
        register(AppIntegrityCallback::class.java)
        register(TextInputCallback::class.java)
    }

    /**
     * Register new Callback Class
     *
     * @param callback The callback Class
     */
    fun register(callback: Class<out Callback>) {
        try {
            callbacks[getType(callback)] = callback
        } catch (e: Exception) {
            error(TAG, e, e.message)
        }
    }

    @Throws(InstantiationException::class, IllegalAccessException::class)
    fun getType(callback: Class<out Callback>): String {
        return callback.getDeclaredConstructor().newInstance().type
    }

    fun getCallbacks(): Map<String, Class<out Callback>> {
        return callbacks
    }

    companion object {
        private val TAG = CallbackFactory::class.java.simpleName

        private val INSTANCE = CallbackFactory()

        /**
         * Returns a cached instance [CallbackFactory]
         *
         * @return instance of [CallbackFactory]
         */
        @JvmStatic fun getInstance(): CallbackFactory {
            return INSTANCE
        }

    }
}
